### PR TITLE
feat(binder): support implicit table function lateral subquery

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Support implicit table function lateral subquery. Resolve https://github.com/risingwavelabs/risingwave/issues/9590
- Related issue https://github.com/risingwavelabs/risingwave/issues/12043

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/src/binder/relation/mod.rs [hunk 1]
```diff
@@ -438,7 +438,10 @@ impl Binder {
                 for_system_time_as_of_proctime,
             } => self.bind_relation_by_name(name, alias, for_system_time_as_of_proctime),
             TableFactor::TableFunction { name, alias, args } => {
-                self.bind_table_function(name, alias, args)
+                self.try_mark_lateral_as_visible();
```
Comment (fuyufjh) [inline src/frontend/src/binder/relation/mod.rs:441]: Do you know why it's named as `try_...`? 🤣

**Review by fuyufjh — APPROVED** [review]
LGTM

**Review by xiangjinwu — COMMENTED** [review]
As mentioned in the issue, `a, b` and `a join b` behaves differently before this PR:
* `a, b` reports `column arr not found`, due to the lack of `try_mark_lateral_as_visible` added in this PR. This is the "support lateral" half of the PR.
* `a cross join b` panics because `arr` was mistakenly accessible in the **current** context (`Unnest($1)` without a child plan node) rather than in lateral or upper context. It is now changed to upper context with `push_context` in `bind_table_function`. This is the "fix panic" half of the PR.

This inspired me to find a trickier case:
```
create table a (a1 int); insert into a values (3), (2);
create table b (b1 varchar); insert into b values ('x'), ('y');
select * from a, b cross join regexp_matches(repeat(b1, a1), '.', 'g');
```
Or similarly:
```
select * from a cross join (b cross join regexp_matches(repeat(b1, a1), '.', 'g'));
```

File: src/frontend/src/binder/relation/mod.rs (hunk 1 above)
Comment (chenzl25) [reply src/frontend/src/binder/relation/mod.rs:441]: Maybe it needs to check if `lateral_context` exists.
```rust
    fn try_mark_lateral_as_visible(&mut self) {
        if let Some(mut ctx) = self.lateral_contexts.pop() {
            ctx.is_visible = true;
            self.lateral_contexts.push(ctx);
        }
    }
```

**Comment by chenzl25** [comment]
> create table a (a1 int); insert into a values (3), (2);
create table b (b1 varchar); insert into b values ('x'), ('y');
select * from a, b cross join regexp_matches(repeat(b1, a1), '.', 'g');


Your cases cause another panic 🥵.

**Comment by chenzl25** [comment]
Even this simpler one.
```sql
create table b (b1 varchar); insert into b values ('x'), ('y');
select * from  b cross join repeat(b1, 2);
```

explain:
```
 LogicalProject { exprs: [b.b1, ] }
 └─LogicalApply { type: Inner, on: true, correlated_id: 1 }
   ├─LogicalScan { table: b, columns: [b1, _row_id] }
   └─LogicalValues { rows: [[Repeat(CorrelatedInputRef { index: 0, correlated_id: 1 }, 2:Int32)]], schema: Schema { fields: [:Varchar] } }
```

CorrelatedInputRef in `LogicalValue` is unexpected. Need to find another way to solve it.

**Comment by chenzl25** [comment]
```sql
explain (trace) select * from a, b cross join c;
```

BTW, this join order is unexpected. I think we should join a and b first instead of join b and c first. cc @xiangjinwu

```
 LogicalProject { exprs: [a.a1, b.b1, c.col] }
 └─LogicalJoin { type: Inner, on: true }
   ├─LogicalScan { table: a, columns: [a1, _row_id] }
   └─LogicalJoin { type: Inner, on: true }
     ├─LogicalScan { table: b, columns: [b1, _row_id] }
     └─LogicalScan { table: c, columns: [col, _row_id] }
```

**Comment by chenzl25** [comment]
Let's merge this PR first. Other unsupported cases mentioned in the comments will be tracked by corresponding issue.

**Pull request merged**